actioncable/app/assets/javascripts/action_cable/connection.coffee

```diff
@@ -89,19 +97,49 @@ class ActionCable.Connection
 
     open: ->
       ActionCable.log("WebSocket onopen event")
-      @disconnected = false
-      @subscriptions.reload()
+      # This is in the case that the browser client opens
+      # the connection regardless of whether a protocol was 
+      # negotiated successfully or not.
+      if @isProtocolSupported()
+        ActionCable.log("WebSocket using '#{@getProtocol()}' subprotocol.")
+        @disconnected = false
+        @subscriptions.reload()
+      else
+        @close()
 
-    close: ->
+    close: (event) ->
       ActionCable.log("WebSocket onclose event")
-      @disconnect()
+
+      # According to https://www.w3.org/TR/websockets/#concept-websocket-close-fail
+      # a sub-protocol mismatch should be treated as a 1006 error, indistinguishable
+      # from any other failure to connect. Not all browsers behave this way:
+
+      # This encompases the behavior of most browsers when it comes to
+      # reacting to a failed protocol negotiation.
+      if event.code is 1000 and event.wasClean and @getProtocol() is ""
+        @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+
+      # This is Chrome's response (which is technically to spec) and looks 
+      # exactly the same as a normal bad disconnect. Thus, we must test to 
+      # see if the server would connect under normal conditions. If so, Chrome reacted
+      # to the failed protocol negotiation, and not something else.
+      else if event.code is 1006 and not event.wasClean and @getProtocol() is ""
+        testSocket = new WebSocket(@consumer.url)
+        testSocket.onopen = =>
+          @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+          testSocket.close()
```

Comment:
So this is all so we can give the app a message saying it was disconnected for good and why.

I'm not a big fan of doing this either, but there isn't a good alternative for Chrome. The reason for all this is that Chrome does not distinguish between a bad disconnect and a protocol mismatch. In either case, it disconnects. For some reason the standard for websockets in browsers wants to disincentivize "probing" a websocket for capabilities (as if you would do this from a web browser!).

Thus, the only way to reliably tell if Chrome disconnected because of your normal bad connection is to create a new connection without protocols to see if Chrome stays connected.

This should only occur in two circumstances:
1. A previous connection attempt on Chrome was unsuccessful at connecting, at which point the server is down anyways.
2. There was a protocol mismatch when the server was up, at which point Chrome is going to disconnect, reconnect again without protocols, see that was the reason, and then disconnect permanently.

The alternative to this is just not upgrading the connection when the server sees that the client wants protocols it doesn't have. I don't think we could get an error code off that.